### PR TITLE
Allow configuring schema for iceberg MV storage tables

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description

Adds storage_schema MV property and
iceberg.materialized-views.storage-schema catalog config.
By default, storage table will continue to created in
the same schema as the MV definition.

> Is this change a fix, improvement, new feature, refactoring, or other?

new feature
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

iceberg connector
> How would you describe this change to a non-technical end user or system administrator?

See description
## Related issues, pull requests, and links

https://github.com/trinodb/trino/issues/12503#issuecomment-1139539874

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Add the ability to configure schema for Iceberg materialized view storage tables. ({issue}`12591`)
```

### CODE REVIEW

**Comment by raunaqmorarka** [comment]
ping @alexjo2144 @findepi

**Comment by alexjo2144** [comment]
Code changes look okay to me. I'm a bit worried about the implications on access control. Granting permissions to uses at the schema level seems like it would be fairly common, and having the data in the same schema enforces they can be accessed by the same users. With the storage table in a different schema, it seems like it'd be easy to accidentally allow users access to the underlying data, who don't have access to the MV.

Maybe I'm being paranoid though.

**Comment by raunaqmorarka** [comment]
> Code changes look okay to me. I'm a bit worried about the implications on access control. Granting permissions to uses at the schema level seems like it would be fairly common, and having the data in the same schema enforces they can be accessed by the same users. With the storage table in a different schema, it seems like it'd be easy to accidentally allow users access to the underlying data, who don't have access to the MV.
> 
> Maybe I'm being paranoid though.

Ideally the storage tables should be hidden and unavailable for direct access from end users. 
With this change the admin would be able to set the storage-schema for MVs in catalog properties and and setup up access controls to deny any direct access to that storage-schema. Some scenarios might require the flexibility to specify different storage-schema for different MVs, that's why it's also available as an MV property. In this case, the access controls could be applied by setting up rules to allow only specific values of storage-schema properties. Although the current access control implementations are not making use if it, the SPI allows inspecting MV properties and implementing rules on them.

**Comment by findepi** [comment]
> worried about the implications on access control.

@kokosing @lukasz-walkiewicz can you please review?

**Comment by raunaqmorarka** [comment]
> > Although the current access control implementations are not making use if it, the SPI allows inspecting MV properties and implementing rules on them
> 
> Is there a plan to do that?

I wasn't planning on implementing the access control check here unless that is a blocker to merging this change. I can file a GH issue to follow-up on it if needed.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewProperties.java
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewProperties.java:38]: Source of truth for table properties is in `IcebergConnector`.
For example, we could switch it to use `TablePropertiesProvider`.

Rename IcebergMaterializedViewProperties to IcebergMaterializedViewAdditionalProperties 
and let the `IcebergConnector` union the two sets in IcebergConnector ctor

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:268]: Who validates that storageSchema does exist?
(previously it was kind of guaranteed by the engine)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:308]: STORAGE_SCHEMA befire STORAGE_TABLE

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:563]: add a test with `storage_schema = 'non_existent'`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:84]: This is going to delete the TESTING_STORAGE_SCHEMA in Glue.
I think we must be using a randomized schema name as the storage schema.

Comment (lukasz-walkiewicz) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:84]: I believe it's already the case
```java
protected static final String TESTING_STORAGE_SCHEMA = "testing_storage_schema_" + randomTableSuffix();
```

**Review by lukasz-walkiewicz — APPROVED** [review]
LGTM % what @findepi already pointed out: having a test for non-existent schema would be great.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java
Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:84]: indeed, i think we ended up with code that i am not smart enough to read :) 

any reason schemaName is an instance field  and TESTING_STORAGE_SCHEMA is a constant, if both are initialized in a pretty same manner (a constant randomized at class load time)?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:268]: We leave it to the user/admin to ensure that the storage schema specified in MV property or catalog config exists. I've added a test with non existent schema to `testStorageSchemaProperty` that verifies the query will fail with Schema not found in exception message.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewProperties.java
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMaterializedViewProperties.java]: Made this change. I've squashed the initial commit now as it looked very awkward to have a preparatory commit with empty properties.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java]: I've made storage schema string consistent with the existing `schemaName` field now

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:588]: Storage table is created before the MV as MV definition records storage table name. So failure to create storage table would ensure that MV is also not created. I've added an assertion about it though.
Can't really check for lack of storage table as the `non_existent` schema doesn't exist and we don't try to create schema ourselves.

**Comment by raunaqmorarka** [comment]
> Needs docs, but looks good to me.
> 
> A follow up issue to think about stricter permissions enforcement would be great too

Added a section in iceberg docs, will file the issue about access controls once this lands.

File: docs/src/main/sphinx/connector/iceberg.rst
Comment (alexjo2144) [inline docs/src/main/sphinx/connector/iceberg.rst:934]: Add this to the `General configuration` table as well, please

Comment (raunaqmorarka) [reply docs/src/main/sphinx/connector/iceberg.rst:940]: done

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:574]: nit: computeScalar

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:569]: maybe use non-empty base table?

**Comment by raunaqmorarka** [comment]
> A follow up issue to think about stricter permissions enforcement would be great too

Filed https://github.com/trinodb/trino/issues/12987

**Pull request merged**
